### ISSUE TITLE
Use "time.After"  may cause a memory leak

### ISSUE THREAD

**Issue by styfstr** [eclipse/paho.mqtt.golang#518]
![image](https://user-images.githubusercontent.com/44523785/125149524-1cc01500-e16c-11eb-81c1-a12672bf6e80.png)

**Comment by MattBrittan** [comment]
Please point out where in this library the memory leak is? 

The comment you pasted does not say that `time.After` causes a memory leak; just that resources are not freed until "the timer fires". This means that a small amount of memory may not be released as rapidly as possible but I think referring to this as a "memory leak" is an exaggeration. Using `time.After` in a tight loop could definitely cause issues but we don't do that (from a quick scan of the code),

There is a potential in that a very high volume publisher may see extra memory usage (due to the call to `time.After(time.Second * 30)` (by default) in `Publish`) which may cause an issue if you have very low resources. You are welcome to submit a PR for this if its impacting you.

**Comment by styfstr** [comment]
Yes, we now use mqtt to communicate in the intranet. The speed of sending and receiving is very fast, about 10-20ms to send data. In the Publish function, time. After will be called frequently, so the memory will continue to grow for a long time to stabilize.I think it's more reasonable to use time.NewTicker here.This can solve unnecessary memory occupation.

**Comment by MattBrittan** [comment]
Why are you preferring ` time.NewTicker` over `time.NewTimer`? (a ticker will fire periodically, a timer will fire once).

**Comment by styfstr** [comment]
Yes, just like you said. For devices with very small memory (for example, my embedded device has only 128M memory), I think this is more reasonable.

**Comment by styfstr** [comment]
Whether time. Newticker or time. Newtimer, resources will be released after calling Stop. It's more reasonable to use time. Newtimer here. Thank you for your correction.

**Comment by jonaz** [comment]
FYI this is the only way to avoid memory leak: 
```
  delay := time.NewTimer(time.Second)
 
  select {
  case <-delay.C:
     // do something after one second.
  case <-ctx.Done():
     // do something when context is finished and stop the timer.
     if !delay.Stop() {
        // if the timer has been stopped then read from the channel.
        <-delay.C
     }    
  }
```

**Comment by jyc6** [comment]
我的也是这个，这个问题造成了内存泄漏，这个文件在哪，怎么修改
